pkg/image/association.go

```diff
@@ -219,146 +230,16 @@ func (as AssociationSet) validate() error {
-		}
-		dirRef := diskLoc.Ref.AsRepository().String()
-		imagePath := filepath.Join(localRoot, dirRef)
-
-		// Verify that the dirRef exists before proceeding
-		if _, err := os.Stat(imagePath); err != nil {
-			errs = append(errs, fmt.Errorf("image %q mapping %q: %v", image, dirRef, err))
-			continue
-		}
-
-		var tagOrID string
-		if diskLoc.Ref.Tag != "" {
-			tagOrID = diskLoc.Ref.Tag
-		} else {
-			tagOrID = diskLoc.Ref.ID
-		}
-
-		if tagOrID == "" {
-			errs = append(errs, &ErrInvalidComponent{image.String(), tagOrID})
-			continue
-		}
-
-		// TODO(estroz): parallelize
-		associations, err := associateImageLayers(image.Ref.String(), localRoot, dirRef, tagOrID, "oc-mirror", image.Category, skipParse)
-		if err != nil {
-			errs = append(errs, err)
-			continue
-		}
-		for _, association := range associations {
-			bundleAssociations.Add(image.Ref.String(), association)
-		}
-	}
-
-	return bundleAssociations, utilerrors.NewAggregate(errs)
-}
-
-func associateImageLayers(image, localRoot, dirRef, tagOrID, defaultTag string, typ ImageType, skipParse func(string) bool) (associations []Association, err error) {
-	if skipParse(image) {
-		return nil, nil
-	}
-
-	manifestPath := filepath.Join(localRoot, filepath.FromSlash(dirRef), "manifests", tagOrID)
-	// TODO(estroz): this file mode checking block is likely only necessary
-	// for the first recursion leaf since image manifest layers always contain id's,
-	// so unroll this component into AssociateImageLayers.
-
-	info, err := os.Lstat(manifestPath)
-	if errors.Is(err, os.ErrNotExist) {
-		return nil, &ErrInvalidComponent{image, tagOrID}
-	} else if err != nil {
-		return nil, err
-	}
-	// Tags are always symlinks due to how `oc` libraries mirror manifest files.
-	id, tag := tagOrID, tagOrID
-	switch m := info.Mode(); {
-	case m&fs.ModeSymlink != 0:
-		// Tag is the file name, so follow the symlink to the layer ID-named file.
-		dst, err := os.Readlink(manifestPath)
-		if err != nil {
-			return nil, fmt.Errorf("error evaluating image tag symlink: %v", err)
-		}
-		id = filepath.Base(dst)
-	case m.IsRegular():
-		// Layer ID is the file name, and no tag exists.
-		tag = defaultTag
-		if defaultTag != "" {
-			// If set, add a subset of the digest to randomize the
-			// tag in the event multiple digests are pulled for the same
-			// image
-			tag = defaultTag + id[7:13]
-			manifestDir := filepath.Dir(manifestPath)
-			symlink := filepath.Join(manifestDir, tag)
-			if err := os.Symlink(info.Name(), symlink); err != nil {
-				return nil, err
-			}
-		}
-	default:
-		return nil, fmt.Errorf("expected symlink or regular file mode, got: %b", m)
-	}
-	manifestBytes, err := ioutil.ReadFile(filepath.Clean(manifestPath))
-	if err != nil {
-		return nil, fmt.Errorf("error reading image manifest file: %v", err)
-	}
+func GetImageFromBlob(as AssociationSet, digest string) string {
+	for imageName, assocs := range as {
+		for _, assoc := range assocs {
+			for _, dgst := range assoc.LayerDigests {
```

Comment:
Why not use `AssociationSet#GetDigest` instead of manually iterating over `AssociaciationSet`'s?